core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/controller/Controller.scala

```diff
@@ -132,6 +155,60 @@ class Controller(
       this.handleControlPayloadWithTryCatch(CLIENT, invocation)
   }
 
+  def acceptRecoveryMessages: Receive = {
+    case recoveryMsg: WorkflowRecoveryMessage =>
+      recoveryMsg.payload match {
+        case UpdateRecoveryStatus(isRecovering) =>
+          logger.info("recovery status for " + recoveryMsg.from + " is " + isRecovering)
+          globalRecoveryManager.markRecoveryStatus(recoveryMsg.from, isRecovering)
+        case ResendOutputTo(vid, ref) =>
+          logger.warn(s"controller should not resend output to " + vid)
+        case NotifyFailedNode(addr) =>
+          if (controllerConfig.supportFaultTolerance) {
+            val deployNodes = availableNodes.filter(_ != self.path.address)
+            if (deployNodes.nonEmpty) {
+              logger.info(
+                "Global Recovery: move all worker from " + addr + " to " + deployNodes.head
+              )
+              val infoIter = workflow.getAllWorkerInfoOfAddress(addr)
+              logger.info("Global Recovery: sent kill signal to workers on failed node")
+              infoIter.foreach { info =>
+                info.ref ! PoisonPill // in case we can still access the worker
+              }
+              logger.info("Global Recovery: triggering worker respawn")
+              infoIter.foreach { info =>
+                val ref = workflow.getWorkerLayer(info.id).recover(info.id, deployNodes.head)
+                logger.info("Global Recovery: respawn " + info.id)
+                val vidSet = infoIter.map(_.id).toSet
+                // wait for some secs to re-send output
+                context.system.scheduler.scheduleOnce(10.seconds) {
+                  logger.info("Global Recovery: triggering upstream resend for " + info.id)
+                  workflow
+                    .getDirectUpstreamWorkers(info.id)
+                    .filter(x => !vidSet.contains(x))
+                    .foreach { vid =>
+                      logger.info("Global Recovery: trigger resend from " + vid + " to " + info.id)
+                      workflow.getWorkerInfo(vid).ref ! ResendOutputTo(info.id, ref)
+                    }
+                }
+              }
+            } else {
+              val error = new RuntimeException(
```

Comment:
add function to throw error. Use it in every place.

Reply:
added this logic to preRestart which is an Akka callback when there is an unhandled exception thrown.